module-system/module-implementations/sov-nft-module/src/offchain.rs

```diff
@@ -0,0 +1,168 @@
+                                  metadata_url = EXCLUDED.metadata_url,\
+                                  supply = EXCLUDED.supply",
+                        &[
+                            &collection_address_str,
+                            &collection_name,
+                            &creator_address_str,
+                            &frozen,
+                            &metadata_url,
+                            &(supply as i64),
+                        ],
+                    );
+                    if let Err(e) = result {
+                        println!("Failed to execute query: {}", e);
+                    }
+                }
+                Err(e) => {
+                    println!("Failed to connect to the database: {}", e);
+                }
+            }
+        } else {
+            println!("Environment variable POSTGRES_CONNECTION_STRING is not set");
+        }
+    })
+}
+
+// CREATE TABLE nft (
+//     collection_address TEXT NOT NULL,
+//     nft_id BIGINT NOT NULL,
+//     metadata_url TEXT,
+//     owner TEXT NOT NULL,
+//     frozen BOOLEAN NOT NULL,
+//     PRIMARY KEY (collection_address, nft_id)
+// );
+
+#[offchain]
+pub fn track_nft<C: sov_modules_api::Context>(nft: &Nft<C>) {
+    let collection_address = nft.get_collection_address().to_string();
+    let nft_id = nft.get_token_id();
+    let owner = nft.get_owner().to_string();
+    let frozen = nft.is_frozen();
+    let metadata_url = nft.get_token_uri();
+    tokio::task::block_in_place(|| {
+        if let Ok(conn_string) = std::env::var("POSTGRES_CONNECTION_STRING") {
+            match postgres::Client::connect(&conn_string, NoTls) {
+                Ok(mut client) => {
+                    let result = client.execute(
+                        "INSERT INTO nfts (\
+                        collection_address, nft_id, metadata_url,\
+                        owner, frozen)\
+                        VALUES ($1, $2, $3, $4, $5)\
+                        ON CONFLICT (collection_address, nft_id)\
+                        DO UPDATE SET metadata_url = EXCLUDED.metadata_url,\
+                                      owner = EXCLUDED.owner,\
+                                      frozen = EXCLUDED.frozen",
+                        &[
+                            &collection_address,
+                            &(nft_id as i64),
+                            &metadata_url,
+                            &owner,
+                            &frozen,
+                        ],
+                    );
+                    if let Err(e) = result {
+                        println!("Failed to execute query: {}", e);
+                    }
+                }
+                Err(e) => {
+                    println!("Failed to connect to the database: {}", e);
+                }
+            }
+        } else {
+            println!("Environment variable POSTGRES_CONNECTION_STRING is not set");
+        }
+    })
+}
```

Comment:
Both of these `track_` functions are long and mostly duplicated. I think we can break them down and then factor some code into a shared helper. 

It looks to me like the two components are: 
1. Build a sql query from the data (depends on the shape of the item we're tracking)
2. Try to execute that query (can take query as a simple argument, identical in both cases)